tools/integration_tests/operations/parallel_dirops_test.go

```diff
@@ -0,0 +1,447 @@
+	"github.com/googlecloudplatform/gcsfuse/v2/tools/integration_tests/util/operations"
+	"github.com/googlecloudplatform/gcsfuse/v2/tools/integration_tests/util/setup"
+	"github.com/stretchr/testify/assert"
+)
+
+// createDirectoryStructureForParallelDiropsTest creates the following files and
+// directory structure.
+// bucket
+//
+//			file1.txt
+//			file2.txt
+//			explicitDir1
+//					file1.txt
+//					file2.txt
+//	    explicitDir2
+//					file1.txt
+func createDirectoryStructureForParallelDiropsTest(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	setup.CleanUpDir(testDir)
+
+	// Create explicitDir1 structure
+	explicitDir1 := path.Join(testDir, "explicitDir1")
+	operations.CreateDirectory(explicitDir1, t)
+	filePath1 := path.Join(explicitDir1, "file1.txt")
+	operations.CreateFileOfSize(5, filePath1, t)
+	filePath2 := path.Join(explicitDir1, "file2.txt")
+	operations.CreateFileOfSize(10, filePath2, t)
+
+	// Create explicitDir2 structure
+	explicitDir2 := path.Join(testDir, "explicitDir2")
+	operations.CreateDirectory(explicitDir2, t)
+	filePath1 = path.Join(explicitDir2, "file1.txt")
+	operations.CreateFileOfSize(11, filePath1, t)
+
+	filePath1 = path.Join(testDir, "file1.txt")
+	operations.CreateFileOfSize(5, filePath1, t)
+	filePath2 = path.Join(testDir, "file2.txt")
+	operations.CreateFileOfSize(3, filePath2, t)
+}
+
+func TestParallelLookUpsForSameFile(t *testing.T) {
+	testDir := setup.SetupTestDirectory(DirForOperationTests)
+	// Create directory structure for testing.
+	createDirectoryStructureForParallelDiropsTest(t)
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(filePath)
+		return fileInfo, err
+	}
+	var stat1, stat2 os.FileInfo
+	var err1, err2 error
+
+	// Parallel lookups of file just under mount.
+	filePath := path.Join(testDir, "file1.txt")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.ErrorIs(t, err1, nil)
+	assert.ErrorIs(t, err2, nil)
+	assert.Equal(t, int64(5), stat1.Size())
+	assert.Equal(t, int64(5), stat2.Size())
+	assert.Contains(t, filePath, stat1.Name())
+	assert.Contains(t, filePath, stat2.Name())
+
+	// Parallel lookups of file under a directory in mount.
+	filePath = path.Join(testDir, "explicitDir1/file2.txt")
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.ErrorIs(t, err1, nil)
+	assert.ErrorIs(t, err2, nil)
+	assert.Equal(t, int64(10), stat1.Size())
+	assert.Equal(t, int64(10), stat2.Size())
+	assert.Contains(t, filePath, stat1.Name())
+	assert.Contains(t, filePath, stat2.Name())
+}
+
+func TestParallelReadDirsForSameDir(t *testing.T) {
```

Comment:
Nit: we are reading different directories also in this test (parent and child explicit directory). Shall we change the name/ create separate tests.

Reply:
done, changed to just TestParallelReadDirs